### PR TITLE
Add the bootloader code to write to program flash

### PR DESCRIPTION
**Pull request by nomis52**

### CODE REVIEW

**Comment by nomis52** [comment]
This also includes the linker script which constrains the bootloader to the first 24k of program flash space.

File: Bootloader/firmware/src/app.c [hunk 1]
```diff
@@ -20,12 +20,39 @@
...
+
+/**
+ * @brief The final address of the user firmware.
+ */
+enum { APP_END_ADDRESS = 0x9d07ffff };
+
+/*
+ * @brief The reset address of the application.
+ */
+enum { APP_RESET_ADDRESS = 0x9d007000 };
```
Comment (peternewman) [inline Bootloader/firmware/src/app.c]: I take it this is supposed to be within the base to end range?

File: Bootloader/firmware/src/app.c [hunk 2]
```diff
@@ -35,20 +62,45 @@ typedef struct {
...
+ * means we could end up with 1-3 bytes that we can't write immediately.
+ *
+ * The solution is to make the incoming DFU buffer slightly bigger so we can
+ * accomodate this, and write them out once we receive the remaining data.
+ */
+static uint8_t DATA_BUFFER[DFU_BLOCK_SIZE + FLASH_WORD_SIZE - 1];
 
 // Helper functions
 // ----------------------------------------------------------------------------
+static inline uint32_t ExtractUInt32(const uint8_t *ptr) {
```
Comment (peternewman) [inline Bootloader/firmware/src/app.c]: Put this in utils.

File: Bootloader/firmware/src/flash.c [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+bool Flash_ErasePage(void *address) {
+  PLIB_NVM_FlashAddressToModify(NVM_ID_0, KVA_TO_PA(address));
+  PerformOperation(PAGE_ERASE_OPERATION);
+
+  while (!PLIB_NVM_FlashWriteCycleHasCompleted(NVM_ID_0)) {
+    ;
+  }
+
+  if (PLIB_NVM_WriteOperationHasTerminated(NVM_ID_0)) {
```
Comment (peternewman) [inline Bootloader/firmware/src/flash.c]: return !PLIB like below.

File: Bootloader/firmware/src/flash.h [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+#include <stdint.h>
+#include <stdbool.h>
+
+/**
+ * @{
+ * @file flash.h
+ * @brief Flash memory operations.
+ *
+ * This file defines constants that are not expected to change.
```
Comment (peternewman) [inline Bootloader/firmware/src/flash.h]: Correct this.

**Comment by peternewman** [comment]
Generally LGTM

File: Bootloader/firmware/src/app.c (hunk 1 above)
Comment (nomis52) [inline Bootloader/firmware/src/app.c]: Yup

File: Bootloader/firmware/src/app.c (hunk 2 above)
Comment (nomis52) [inline Bootloader/firmware/src/app.c]: There is no utils. Almost all the bootloader code is in a single file. If I need to use it elsewhere I'll move it out.

File: Bootloader/firmware/src/flash.c (hunk 3 above)
Comment (nomis52) [inline Bootloader/firmware/src/flash.c]: Fixed in the next change

File: Bootloader/firmware/src/flash.h (hunk 4 above)
Comment (nomis52) [inline Bootloader/firmware/src/flash.h]: Done

**Pull request merged**